feat(newsletter, contact) strengthen types and remove generics

hanmindev opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/develop/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Fixes: N/A
Touches #1212

## What is the new behavior?


## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->
Generics removed from:
contact
- DaffContactServiceInterface
- DaffContactSubmit
- DaffContactRetry
- DaffContactActions

newsletter
- DaffNewsletterServiceInterface
- DaffNewsletterSubscribe
- DaffNewsletterRetry
- DaffNewsletterActions
- DaffNewsletterEffects

DaffNewsletterUnion removed. Replaced with more appropriate interfaces for each case

## Other information

damienwebdev commented on libs/newsletter/driver/hubspot/src/newsletter.service.spec.ts:
```diff
@@ -1,42 +1,72 @@
-      const expected = cold('--b', { b: { test: '123' }});
-
-      expect(newsletterService.send(payload)).toBeObservable(expected);
+  it('should allow a developer to configure and send newsletter subscription requests to the HubspotForms API', () => {
```
We need a test covering the string type of the arg.

damienwebdev commented on libs/newsletter/driver/hubspot/src/newsletter.service.ts:
```diff
@@ -5,17 +5,31 @@ import {
-export class DaffNewsletterHubspotService implements DaffNewsletterServiceInterface<DaffNewsletterUnion, any> {
+export class DaffNewsletterHubspotService implements DaffNewsletterServiceInterface {
 
   constructor(@Inject(DAFF_NEWSLETTER_HUBSPOT_FORMS_TOKEN) private hubspotService: DaffHubspotFormsService) {}
```
This should be the `DaffHubspotFormsInterface` not `DaffHubspotFormsService`

damienwebdev commented on libs/newsletter/driver/hubspot/src/newsletter.service.spec.ts:
```diff
@@ -1,42 +1,72 @@
 
   beforeEach(() => {
     TestBed.configureTestingModule({
+      imports: [
```
This should not use the `HttpClientTestingModule`. You should either expose a mock of the driver of the `@daffodil/driver/hubspot/testing` package or just make a simple mock here.

damienwebdev commented on libs/newsletter/src/models/newsletter_storage.ts:
```diff
@@ -0,0 +1,6 @@
+/**
+ * A type for the storage of a newsletter subscription.
+ */
+export type DaffNewsletterStorage = {
```
This looks like its only used in the `in-memory` service, let's move it there.

damienwebdev commented on libs/newsletter/src/models/newsletter_response.ts:
```diff
@@ -0,0 +1,6 @@
+/**
+ * A type for the response of a newsletter subscription.
+ */
+export type DaffNewsletterResponse = string | {
```
This should only be a an object, not a string.

damienwebdev commented on libs/core/state/src/errors/public_api.ts:
```diff
@@ -2,5 +2,5 @@ export { DaffStateError } from './state-error.interface';
 export { DaffErrorable } from './errorable.type';
 export {
   daffTransformErrorToStateError,
```
This can be reverted.

damienwebdev commented on libs/contact/driver/hubspot/src/config/contact-config.interface.ts:
```diff
@@ -2,6 +2,9 @@ import { InjectionToken } from '@angular/core';
 import { DaffHubspotConfig } from '@daffodil/driver/hubspot';
 
+/**
+ * The injection token that holds the configuration of the DaffContact feature.
```
This doc looks incorrect. This configuration token is specific to the Hubspot driver.

damienwebdev commented on libs/contact/driver/testing/src/contact.service.ts:
```diff
@@ -1,21 +1,24 @@
-  send(payload: DaffContactUnion): Observable<any>{
-    return of('Success').pipe(delay(10));
+export class DaffTestingContactService implements DaffContactServiceInterface {
+  send(payload: DaffContactRequest): Observable<DaffContactResponse> {
```
Changing this should have failed a test, can you add one?

damienwebdev commented on libs/contact/state/src/actions/contact.actions.ts:
```diff
@@ -11,16 +12,16 @@ export enum DaffContactActionTypes {
 }
 
-export class DaffContactSubmit<T> implements Action {
+export class DaffContactSubmit implements Action {
```
Can you document this action?

damienwebdev commented on libs/contact/state/src/actions/contact.actions.ts:
```diff
@@ -11,16 +12,16 @@ export enum DaffContactActionTypes {
 }
 
-export class DaffContactRetry<T> implements Action {
+export class DaffContactRetry implements Action {
```
Can you document this action?

damienwebdev commented on libs/contact/state/src/actions/contact.actions.ts:
```diff
@@ -11,16 +12,16 @@ export enum DaffContactActionTypes {
-  constructor(public payload: T) {}
+  constructor(public payload: DaffContactRequest) {}
 }
 export class DaffContactFailedSubmit implements Action {
```
Can you document all of these actions?

damienwebdev commented on libs/driver/hubspot/src/hubspot-forms.provider.ts:
```diff
@@ -4,12 +4,12 @@ import { Router } from '@angular/router';
-
+import { DaffHubspotFormsInterface } from './models/forms';
 
 export const daffHubspotFormsServiceFactory = (
```
Can you document this function?

damienwebdev commented on libs/driver/hubspot/src/hubspot-forms.service.ts:
```diff
@@ -4,11 +4,12 @@ import { Router } from '@angular/router';
 import { jsonBuilder } from './transformers/json-builder';
 
-export class DaffHubspotFormsService {
+export class DaffHubspotFormsService implements DaffHubspotFormsInterface {
```
Can you document this service?

damienwebdev commented on libs/driver/hubspot/src/models/forms.ts:
```diff
@@ -0,0 +1,7 @@
+
+import { HubspotResponse } from './hubspot-response';
+
+export interface DaffHubspotFormsInterface {
```
Needs docs

damienwebdev commented on libs/driver/hubspot/src/models/forms.ts:
```diff
@@ -0,0 +1,7 @@
+import { HubspotResponse } from './hubspot-response';
+
+export interface DaffHubspotFormsInterface {
+  submit(payload: Record<string, any>): Observable<HubspotResponse>;
```
Needs docs

damienwebdev commented on libs/driver/hubspot/testing/src/factories/hubspot-response.factory.ts:
```diff
@@ -0,0 +1,31 @@
+@Injectable({
+  providedIn: 'root',
+})
+export class DaffHubspotResponseFactory extends DaffModelFactory<HubspotResponse>{
```
Needs docs

damienwebdev commented on libs/newsletter/driver/hubspot/src/newsletter.service.ts:
```diff
@@ -3,19 +3,34 @@ import {
 
 @Injectable()
-export class DaffNewsletterHubspotService implements DaffNewsletterServiceInterface<DaffNewsletterUnion, any> {
+export class DaffNewsletterHubspotService implements DaffNewsletterServiceInterface {
```
Needs docs, probably `inheritdoc`

damienwebdev commented on libs/newsletter/driver/hubspot/src/token/hubspot-forms.token.ts:
```diff
@@ -8,13 +8,16 @@ import { Title } from '@angular/platform-browser';
+/**
+ * A token for combining the inputs to the HubspotFormsServiceFactory into a single dependency.
+ */
+export const DAFF_NEWSLETTER_HUBSPOT_FORMS_TOKEN: InjectionToken<DaffHubspotFormsInterface> = new InjectionToken<DaffHubspotFormsInterface>('DAFF_NEWSLETTER_HUBSPOT_FORMS_TOKEN',
```
Why is this doc different than the `contact` version of the same token? Can you give an example usage in the doc here to clarify how one would use it?

damienwebdev commented on libs/newsletter/driver/in-memory/src/backend/newsletter.service.ts:
```diff
@@ -5,13 +5,13 @@ import {
 @Injectable({
   providedIn: 'root',
 })
 export class DaffInMemoryBackendNewsletterService implements InMemoryDbService {
```
Needs docs.

damienwebdev commented on libs/newsletter/driver/in-memory/src/drivers/newsletter.service.ts:
```diff
@@ -13,19 +16,19 @@ import { DaffNewsletterServiceInterface } from '@daffodil/newsletter/driver';
   providedIn: 'root',
 })
-export class DaffInMemoryNewsletterService implements DaffNewsletterServiceInterface<DaffNewsletterUnion, DaffNewsletterUnion>{
+export class DaffInMemoryNewsletterService implements DaffNewsletterServiceInterface {
```
Needs `@inheritdoc`

damienwebdev commented on libs/newsletter/driver/in-memory/src/models/newsletter_storage.ts:
```diff
@@ -0,0 +1,6 @@
+/**
+ * A type for the storage of a newsletter subscription.
+ */
+export interface DaffNewsletterStorage {
```
Can you clarify the purposes of this?

damienwebdev commented on libs/newsletter/driver/src/interfaces/newsletter-service.interface.ts:
```diff
@@ -1,10 +1,19 @@
+/**
+ * The interface responsible for sending newsletter submissions.
+ */
+export interface DaffNewsletterServiceInterface {
```
Can you improve this doc a bit? It would be nice to explain the method, its params, and what other docs tokens would be nice to see for further info. You can look for `@link` or `@see` in the codebase for examples.

damienwebdev commented on libs/newsletter/driver/testing/src/drivers/newsletter.service.ts:
```diff
@@ -5,15 +5,18 @@ import {
+} from '@daffodil/newsletter';
 import { DaffNewsletterServiceInterface } from '@daffodil/newsletter/driver';
 
 @Injectable({
```
Needs `@inheritdoc`

damienwebdev commented:
@griest024 needs rebase

griest024 committed:
feat(driver): strongly type hubspot forms service

griest024 committed:
feat(newsletter)!: remove unecessary generics and improve tests

BREAKING CHANGE: `DaffNewsletterState` now uses `DaffOperationState`. Generics have been removed and the models have been narrowed.

griest024 committed:
feat(contact)!: remove unecessary generics and improve tests

BREAKING CHANGE: `DaffContactState` now uses `DaffOperationState`. Generics have been removed and the models have been narrowed.

griest024 committed:
fix(demo): newsletter breaking changes
